bdns/servers.go

```diff
@@ -113,23 +124,79 @@ type dynamicProvider struct {
 	updateCounter *prometheus.CounterVec
 }
 
+// resolveDNSAuthority resolves the DNS authority to use for resolution of other
+// DNS backends. The DNS authority can be specified as a hostname or IP address,
+// with or without a port. If the authority is specified as a hostname it will
+// be resolved via the system DNS. If the authority is specified as an IP
+// address it will be used directly, defaulting to port 53 if no port is
+// specified.
+func resolveDNSAuthority(d string) (string, error) {
+	host, port, err := net.SplitHostPort(d)
+	if err != nil {
+		// Assume host with no port specified, default port to 53.
+		host = d
+		port = "53"
+	} else {
+		// Ensure the `port` portion of `address` is a valid port.
+		portNum, err := strconv.Atoi(port)
+		if err != nil {
+			return "", fmt.Errorf("parsing port number: %s", err)
+		}
+		if portNum <= 0 || portNum > 65535 {
+			return "", errors.New("port must be an integer between 0 - 65535")
+		}
+	}
+
+	if net.ParseIP(host) != nil {
+		return net.JoinHostPort(host, port), nil
+	}
+	ips, err := net.LookupIP(host)
+	if err != nil {
+		return "", fmt.Errorf("during A/AAAA lookup of %q: %s", d, err)
+	}
+	if len(ips) <= 0 {
+		return "", fmt.Errorf("A/AAAA lookup for %q returned 0 results", d)
+	}
+	return net.JoinHostPort(ips[0].String(), port), nil
+}
+
 var _ ServerProvider = &dynamicProvider{}
 
 // StartDynamicProvider constructs a new dynamicProvider and starts its
 // auto-update goroutine. The auto-update process queries DNS for SRV records
 // at refresh intervals and uses the resulting IP/port combos to populate the
 // list returned by Addrs. The update process ignores the Priority and Weight
-// attributes of the SRV records. The given server name should be a full domain
-// name like `example.com`, which will result in SRV queries for `_dns._udp.example.com`.
-func StartDynamicProvider(server string, refresh time.Duration) (*dynamicProvider, error) {
-	if server == "" {
-		return nil, fmt.Errorf("no DNS domain name provided")
+// attributes of the SRV records.
+func StartDynamicProvider(c *cmd.DNSProvider, refresh time.Duration) (*dynamicProvider, error) {
+	if c.SRVLookup.Domain == "" {
+		return nil, fmt.Errorf("'domain' cannot be empty")
 	}
+
+	service := c.SRVLookup.Service
+	if service == "" {
+		// Default to "dns" if no service is specified. This is the default
+		// service name for DNS servers.
+		service = "dns"
+	}
+
+	// TODO(#6868): Make dnsAuthority required once 'dnsResolver' is removed
+	// from the boulder-va JSON config in favor of 'dnsProvider'.
+	dnsAuthority := c.DNSAuthority
+	if dnsAuthority != "" {
+		resolved, err := resolveDNSAuthority(dnsAuthority)
```

Comment:
One subtle thing I notice: here we resolve the dnsAuthority hostname (e.g. `consule.service.consul`) once at startup. That's probably fine; I'm not actually sure what the gRPC resolver does (is Build called multiple times?), but we should aim to match it.

Reply:
Ah, after looking at how it's handled in gRPC I notice that we actually don't need to resolve the dnsAuthority here. The dialer that we're constructing for our net.Resolver will use a net.DefaultResolver to resolve any hostname passed at Dial time. This is actually something we already rely on in our custom gRPC SRV resolver and I think this makes the whole implementation less brittle as you won't have to restart the process to re-resolve the DNS hostname.